lib/stages/BeamBufferSort.cpp

```diff
@@ -0,0 +1,295 @@
+	buf_start_f += buf_nchan;
+    }
+    
+}
+
+BeamBufferSort::~BeamBufferSort() {}
+
+void BeamBufferSort::make_freq_meta(FreqIDBeamMetadata* out_meta, uint32_t freq_id,\ 
+    uint64_t fpga_seq_start, timespec ctime, uint64_t stream_id, dset_id_t dataset_id,\
+    uint32_t beam_number, float ra, float dec, uint32_t scaling){
+    out_meta -> fpga_seq_start = fpga_seq_start;
+    out_meta -> ctime = ctime;
+    out_meta -> stream_id.id = stream_id;
+    out_meta -> dataset_id = dataset_id;
+    out_meta -> beam_number = beam_number;
+    out_meta -> ra = ra;
+    out_meta -> dec = dec;
+    out_meta -> scaling = scaling;
+    out_meta -> frequency_bin = freq_id;
+}
+
+void BeamBufferSort::nonfreq_meta_2_freq_meta(BeamMetadata* nonfreq_meta,\
+    FreqIDBeamMetadata* freq_meta, uint32_t freq_id){
+    make_freq_meta(freq_meta, freq_id, nonfreq_meta -> fpga_seq_start,\
+        nonfreq_meta-> ctime, nonfreq_meta -> stream_id.id, nonfreq_meta -> dataset_id,\
+	nonfreq_meta -> beam_number, nonfreq_meta -> ra, nonfreq_meta -> dec,\
+	nonfreq_meta -> scaling);
+}
+
+void BeamBufferSort::fill_empty_frame(uint32_t time_idx, uint32_t freq_idx, 
+    uint64_t fpga_seq_start0, timespec ctime0, uint32_t beam_number, double ra, 
+    double dec, double scaling){
+    
+    uint32_t fill_pos;
+    uint64_t fake_fpga_seq_start;
+    uint32_t fake_freq_bin;     // For metadata in the empty frame
+    timespec fake_ctime;
+
+    fill_pos = freq_idx * sub_frame_size;
+    fake_fpga_seq_start = fpga_seq_start0 + time_idx;
+    fake_ctime = ctime0;
+    add_nsec(fake_ctime, (long)(time_idx * subframe_time_nsec));
+    fake_freq_bin = freq_idx;
+    make_freq_meta((FreqIDBeamMetadata*) &sort_queue[time_idx][fill_pos], fake_freq_bin,
+        fake_fpga_seq_start, fake_ctime, 0, dset_id_t::null, beam_number, ra, dec, scaling);
+    memset(&sort_queue[time_idx][fill_pos + FreqIDBeamMeta_size], 0, samples_per_data_set);
+    queue_status[time_idx][freq_idx] = 1;
+}
+
+void BeamBufferSort::main_thread(){
+    uint32_t freq_bin;
+    int time_offset;
+    uint32_t start_pos;
+    uint64_t frame_nr;
+    uint8_t* output;
+    uint32_t dump_nframe;
+    uint32_t dump_start;
+    uint32_t dump_size;
+    uint64_t fpga_seq_start;
+    uint64_t fpga_seq_start0;
+    uint64_t frame0 = 0;
+    uint32_t beam_number;
+    double ra;
+    double dec;
+    double scaling;
+
+    timespec ctime0;
+    
+    FreqIDBeamMetadata* in_metadata = nullptr;
+    BeamMetadata* non_freq_metadata = nullptr;
+
+    frameID input_frame_id(in_buf); // Input frame id
+    std::vector<frameID> output_frame_ids; 
+    for (auto& out_buf : out_bufs){
+        output_frame_ids.push_back(frameID(out_buf));
+    }
+    
+    while (!stop_thread) {
+	// Receive a new frame
+        uint8_t* input = wait_for_full_frame(in_buf, unique_name.c_str(), input_frame_id);
+	if (input == nullptr)
+            break;
+        
+	// Get the metadata for each frame.
+        if (has_freq_bin) {
```

Comment:
It's likely more flexible to setup this variable as a string like `metadata_type`, and then you can check `metadata_type == "BeamMetadata"`, etc.  It's a little more self-descriptive than having a boolean, and more flexible for other types.  Ideally the buffer would tell you what metadata it has and you would select on that, but it currently doesn't support that.